### PR TITLE
Headerbar implementation into GTG!

### PR DESCRIPTION
**Pull request by jakubbrindza**
This pull request introduces HeaderBar into the GTG app.

I have included the most essential buttons and functionality which had to be removed from the menu and toolbar. Now they provide the very basic functionality so as to allow normal, comfortable use.

Other:
 - refurbished keyboard shortcuts in order to preserve the ease of use as before
 - some functions which may not be currently in use have been preserved so as to allow their use in a short period of time when I implement stack switcher or other features.
 - over the course of the upcoming months, when the design polishes, the text buttons will be probably changed by representative icons.

Please, note that the centred three buttons will be redesigned into a StackSwitcher shortly, now they are being included only as normal buttons due to the functionality constraint.

screen shot 2015-05-26 at 21 10 41

### CODE REVIEW

File: src/liblarch [hunk 1]
```diff
@@ -0,0 +1 @@
+Subproject commit b8736982bd225fae5e3271fe653e85c2534bb99e
```
Comment (izidormatusov) [inline src/liblarch]: Please, remove this file. It should have been in your .gitignore to prevent you committing it.

File: src/pip-delete-this-directory.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+This file is placed here by pip to indicate the source was put
```
Comment (izidormatusov) [inline src/pip-delete-this-directory.txt]: Please, remove this file. It should have been in your .gitignore to prevent you committing it.

File: GTG/gtk/browser/taskbrowser.ui [hunk 3]
```diff
@@ -8,529 +8,176 @@
...
               </packing>
             </child>
           </object>
-          <packing>
-            <property name="expand">False</property>
-            <property name="fill">False</property>
-            <property name="position">1</property>
-          </packing>
         </child>
+    <child>
```
Comment (izidormatusov) [inline GTG/gtk/browser/taskbrowser.ui]: please, indent this file properly. I'm sure your editor might have support for doing it automatically

File: GTG/gtk/browser/browser.py [hunk 4]
```diff
@@ -726,7 +707,7 @@ def hide_closed_pane(self):
         #     self.vtree_panes['closed'].set_model(None)
         #     del self.vtree_panes['closed']
         self.remove_page_from_accessory_notebook(self.closed_pane)
-        self.builder.get_object("view_closed").set_active(False)
+        #self.builder.get_object("view_closed").set_active(False)
         self.config.set('closed_task_pane', False)
 
     def on_toolbar_toggled(self, widget):
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: this method can be removed as is not used anywhere

File: GTG/gtk/browser/browser.py [hunk 5]
```diff
@@ -210,22 +195,6 @@ def init_tags_sidebar(self):
...
-        sets tooltips for widgets on toolbars
-        """
-        self.donebutton.set_tooltip_text(GnomeConfig.MARK_DONE_TOOLTIP)
-        self.editbutton.set_tooltip_text(GnomeConfig.EDIT_TOOLTIP)
-        self.dismissbutton.set_tooltip_text(GnomeConfig.MARK_DISMISS_TOOLTIP)
-        self.newtask.set_tooltip_text(GnomeConfig.NEW_TASK_TOOLTIP)
-        self.newsubtask.set_tooltip_text(GnomeConfig.NEW_SUBTASK_TOOLTIP)
-        self.toggle_workview.set_tooltip_text(
-            GnomeConfig.WORKVIEW_TOGGLE_TOOLTIP)
-        self.quickadd_entry.set_tooltip_text(
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: Can you bring back this tooltip and set it directly in taskbrowser.ui? I don't think it's dynamically set anywhere.

File: GTG/gtk/browser/browser.py [hunk 6]
```diff
@@ -210,22 +195,6 @@ def init_tags_sidebar(self):
...
-            GnomeConfig.QUICKADD_ENTRY_TOOLTIP)
-        quickadd_icon_tooltip = GnomeConfig.QUICKADD_ICON_TOOLTIP
-        self.quickadd_entry.set_icon_tooltip_text(1,
-                                                  quickadd_icon_tooltip)
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: Can you bring back this tooltip and set it directly in taskbrowser.ui? I don't think it's dynamically set anywhere.

File: GTG/gtk/browser/browser.py [hunk 7]
```diff
@@ -382,19 +357,19 @@ def _init_accelerators(self):
...
-        self._add_accelerator_for_widget(agr, "view_sidebar", "F9")
-        self._add_accelerator_for_widget(agr, "file_quit", "<Control>q")
-        self._add_accelerator_for_widget(agr, "edit_undo", "<Control>z")
-        self._add_accelerator_for_widget(agr, "edit_redo", "<Control>y")
-        self._add_accelerator_for_widget(agr, "new_task_mi", "<Control>n")
-        self._add_accelerator_for_widget(agr, "new_subtask_mi",
+        self._add_accelerator_for_widget(agr, "tags", "F9")
+        #self._add_accelerator_for_widget(agr, "file_quit", "<Control>q")
+        #self._add_accelerator_for_widget(agr, "edit_undo", "<Control>z")
+        #self._add_accelerator_for_widget(agr, "edit_redo", "<Control>y")
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: Please remove shortcuts for undo and redo. They're not supported (yet)

File: GTG/gtk/browser/browser.py [hunk 8]
```diff
@@ -462,10 +446,10 @@ def restore_state_from_conf(self):
 
         tag_pane = self.config.get("tag_pane")
         if not tag_pane:
-            self.builder.get_object("view_sidebar").set_active(False)
+            #self.builder.get_object("view_sidebar").set_active(False)
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: You can use the same logic for Tag button:

```python
self.builder.get_object("tags").set_active(False)
```

File: GTG/gtk/browser/browser.py [hunk 9]
```diff
@@ -462,10 +446,10 @@ def restore_state_from_conf(self):
...
             self.sidebar.hide()
         else:
-            self.builder.get_object("view_sidebar").set_active(True)
+            #self.builder.get_object("view_sidebar").set_active(True)
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: Same as above

File: GTG/gtk/browser/browser.py [hunk 10]
```diff
@@ -487,19 +471,19 @@ def restore_state_from_conf(self):
...
-            self.builder.get_object("view_quickadd").set_active(True)
-        else:
-            self.quickadd_pane.hide()
-            self.builder.get_object("view_quickadd").set_active(False)
+        # quickadd_pane = self.config.get("quick_add")
+        # if quickadd_pane:
+        #     self.builder.get_object("view_quickadd").set_active(True)
+        # else:
+        #     self.quickadd_pane.hide()
+        #     self.builder.get_object("view_quickadd").set_active(False)
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: I think we can get rid of both blocks. 

Also, would you mind to merge back upstream changes in your branch? I think there might be a merge conflict. 

After you do that, please add todo to file GTG/core/config.py to remove toolbar and quick_add options from configuration (as we probably won't need them by release time)

File: GTG/gtk/browser/browser.py [hunk 11]
```diff
@@ -569,8 +553,8 @@ def set_view(self, viewname):
             workview = True
         else:
             raise Exception('Cannot set the view %s' % viewname)
-        self.menu_view_workview.set_active(workview)
-        self.toggle_workview.set_active(workview)
+        # self.menu_view_workview.set_active(workview)
+        # self.toggle_workview.set_active(workview)
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: remove those lines

File: GTG/gtk/browser/browser.py [hunk 12]
```diff
@@ -656,13 +640,14 @@ def on_workview_toggled(self, widget):
         self.do_toggle_workview()
 
     def on_sidebar_toggled(self, widget):
-        view_sidebar = self.builder.get_object("view_sidebar")
+        pass
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: drop `pass`

File: GTG/gtk/browser/browser.py [hunk 13]
```diff
@@ -656,13 +640,14 @@ def on_workview_toggled(self, widget):
...
+        #view_sidebar = self.builder.get_object("view_sidebar")
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: can you repurpose the code to work with your new `tags` button? uncomment it and change view_sidebar to tags

File: GTG/gtk/browser/browser.py [hunk 14]
```diff
@@ -726,7 +707,7 @@ def hide_closed_pane(self):
         #     self.vtree_panes['closed'].set_model(None)
         #     del self.vtree_panes['closed']
         self.remove_page_from_accessory_notebook(self.closed_pane)
-        self.builder.get_object("view_closed").set_active(False)
+        #self.builder.get_object("view_closed").set_active(False)
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: Drop the line

File: GTG/gtk/browser/taskbrowser.ui [hunk 15]
```diff
@@ -8,529 +8,176 @@
...
+                    <property name="label">Settings</property>
+                    <property name="sensitive">True</property>
+                    <property name="tooltip_text" translatable="yes">Settings/Preferences
+                    </property>
+                    <signal name="clicked" handler="on_preferences_activate" swapped="no"/>
+                    <property name="valign">center</property>
                   </object>
+                  <style>
+                    <class name="text-button"/>
+                  </style>
```
Comment (izidormatusov) [inline GTG/gtk/browser/taskbrowser.ui]: you miss here `packing` information

File: GTG/gtk/browser/taskbrowser.ui [hunk 16]
```diff
@@ -8,529 +8,176 @@
...
                 </child>
+              </object>
+            </child>
+            <child>
+              <object class="GtkButton" id="search">
+                <property name="visible">True</property>
+                <property name="sensitive">True</property>
+                <property name="tooltip_text" translatable="yes">Searching thing</property>
```
Comment (izidormatusov) [inline GTG/gtk/browser/taskbrowser.ui]: Can you fix the tooltip_text. We might forget about it and user might be confused ;)

File: GTG/gtk/browser/taskbrowser.ui [hunk 17]
```diff
@@ -8,529 +8,176 @@
...
+                  <property name="visible">True</property>
+                  <property name="label">Plugins</property>
+                  <property name="sensitive">True</property>
+                  <property name="tooltip_text" translatable="yes">Add plugins</property>
+                  <signal name="clicked" handler="on_edit_plugins_activate" swapped="no"/>
+                  <property name="valign">center</property>
+                </object>
+                <style>
+                  <class name="text-button"/>
+                </style>
```
Comment (izidormatusov) [inline GTG/gtk/browser/taskbrowser.ui]: you are missing `packing` information here

File: GTG/gtk/browser/taskbrowser.ui [hunk 18]
```diff
@@ -8,529 +8,176 @@
...
-                    <property name="expand">False</property>
-                    <property name="homogeneous">True</property>
-                  </packing>
                 </child>
               </object>
               <packing>
-                <property name="expand">False</property>
-                <property name="fill">True</property>
-                <property name="position">0</property>
+                <property name="pack_type">end</property>
```
Comment (izidormatusov) [inline GTG/gtk/browser/taskbrowser.ui]: I don't think that you intended to put the search_icon to the very right side of the pane ;)

**Comment by jakubbrindza** [comment]
@izidormatusov please take a look, all your comments should be resolved by now and the PR ready for merging.
Thanks

File: GTG/gtk/browser/browser.py [hunk 19]
```diff
@@ -713,7 +683,7 @@ def show_closed_pane(self):
             return
 
         self.add_page_to_accessory_notebook("Closed", self.closed_pane)
-        self.builder.get_object("view_closed").set_active(True)
+        #self.builder.get_object("view_closed").set_active(True)
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: Can you remove this line? No need for code that was commented out

File: GTG/core/config.py [hunk 20]
```diff
@@ -21,6 +21,9 @@
 Classes responsible for handling user configuration
 """
 
+# TODO: Remove toolbar and quick_add options from configuration
+#       They won't be used in GTG 0.4
+
```
Comment (izidormatusov) [inline GTG/core/config.py]: Can you put this on top of DEFAULTS dict as that's the place that need to be modified?

**Comment by tkdchen** [comment]
Such several buttons put on the HeaderBar is not extensible for the feature growth in the future.

In my opinion, "New Task" is proper to put on the HeaderBar as button, and all others should be grouped  into menuitems.

**Comment by izidormatusov** [comment]
@tkdchen this is just a stepping stone in @jakubbrindza 's [GSoC project](https://wiki.gnome.org/Outreach/SummerOfCode/2015/Projects/JakubBrindza_GTGRedesign) where he will slowly implement new, more [GNOME3 design](https://wiki.gnome.org/Apps/GTG/Design). Those buttons will be removed eventually.

Jakub and I are experimenting with the different approach to GSoC. Instead of Jakub working on his private branch and making one massive PR at the end of summer that will take another 6 months to merge and resolve all issues, Jakub is going to do small, understandable changes over the summer.

**TL;DR** The buttons will be removed in one of next PRs.

**Comment by jakubbrindza** [comment]
@tkdchen thank you for your commentary, it is good to have someone else apart from me and @izidormatusov to point out facts about the work being done.

However, I agree with Izidor. We decided to move towards GTG 0.4 by doing several small steps rather than a huge leap. In a short period of time, I will deliver stack switcher into GTG which will tidy up the headerbar a bit and when we implement app menu afterwards, majority of the buttons you can see in headerbar now will be moved there (Settings, Plugins, Sync. Services, Tags Sidebar).

That will hopefully lead to a headerbar with less clutter, as you well suggested :)

File: GTG/gtk/browser/browser.py [hunk 21]
```diff
@@ -212,20 +200,10 @@ def init_tags_sidebar(self):
 
     def _init_toolbar_tooltips(self):
         """
-        sets tooltips for widgets on toolbars
+        Sets tooltips for widgets which cannot be setup in .ui yets
```
Comment (izidormatusov) [inline GTG/gtk/browser/browser.py]: typo, s/yets/yet/

**Comment by tkdchen** [comment]
Thanks for sharing this background.
